<section id='exception-test'>
  <div class='grid'>
    <div class='text'>
      <h2>
        <a href='#exception-test' class='anchor'>¶</a>
        Тестируем исключения
      </h2>
    </div>
  </div>
  <div class='grid'>
    <div class='text'>
      <p>Теперь продумаем ситуацию, когда пользвователь пытается пойти в уже занятую клетку.</p>
      <p>Такую ситуацию можно обыграть по-разному, но мы будем в таких случаях выбрасывать исключения.</p>
      <p>Чтобы протестировать подобную ситуацию, нам нужно сделать ход в какую-то клетку, а затем сделать ход в неё ещё раз.</p>
      <p>Подготовку мы делаем именно в этом тесте. Нельзя где-то в начале создать новую игру, в первом тесте её изменить, а во втором использовать эти изменения, как начальное состояние.</p>
      <p>Каждый тест должен быть независим друг от друга. В правильно структурированном наборе тестов, ничего не сломается, если вы какой-то тест выкинете или поменяете их местами.</p>
    </div>

    <div class='code'>
<pre data-start='32'><div class='code-filename'>test/game.js</div><code class='language-js line-numbers'>it('Throws an exception if user moves in taken cell', () => {
  const x = 2, y = 2

  game.acceptUserMove(x, y)
  const func = game.acceptUserMove.bind(game, x, y)
  
  expect(func).to.throw('cell is already taken')
})</code></pre>
    </div>
  </div>
</section>